README.md
# coolharm3d

doc/CHANGES
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@  										  @@
@@  THIS FILE DESCRIBES THE STEPS TAKEN IN TAKING GAMMIE'S HAM3D CODE AND	  @@
@@  MAKING IT A RELATIVISTIC MHD CODE WITH MPI SUPPORT CALLED HARM3D.		  @@
@@  Please see "README" for a description of the algorithm and how	          @@
@@  to use it.									  @@
@@  										  @@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@





################################
Sat Feb 18 21:21:54 CST 2006
-------------------------------

-- this directory is my development directory in which I
   am altering starting from ham3d

-----------------------------------------------------

PRELIMINARY HAM3D CHANGES:
-- my goal is to change some of the variable names before
	making the split to ham3d_mpi


    N2->N3
    N1->N2
    N0->N1
	-- searching for others:
		prompt> cat *.c *.h| awk '/0+.*1+.*2+/' | less

	-- doing the same for L[0-2], U[0-2], B[0-2], dx[0-2], X[0-2], p[0-2], ek[0-2],
                              , v[0-2], b[0-2], [0-2]-face, j[0-2], w[0-2], c[0-2],
				eb[0-2] (but don't do since b[0-2] will handle it)


-- we can do this all at once, working on the wild side here...:

foreach n ( N L U B dx X p ek v b j w c )
foreach? sed -i s/${n}2/${n}3/ *.c *.h
foreach? sed -i s/${n}1/${n}2/ *.c *.h
foreach? sed -i s/${n}0/${n}1/ *.c *.h
foreach? end

prompt> sed -i s/2-face/3-face/ *.c *.h
prompt> sed -i s/1-face/2-face/ *.c *.h
prompt> sed -i s/0-face/1-face/ *.c *.h


-- it seems to have worked, I'll keep an eye out for any stray ones...


#############################
Wed Sep 13 12:33:27 EDT 2006
-------------------------------

-- actually it didn't work since I failed to use the 'g' option with
   the search/replace, so the pattern was substituted only once per
   matching line (and there were lines with multiple matches).

-- since I don't remember making any other changes to the code,
    I'm going to start with the original version and then redo the
    sed commands:
    -- the new sed commands are:

foreach n ( N L U B dx X p ek v b j w c )
foreach? sed -i s/${n}2/${n}3/g *.c *.h
foreach? sed -i s/${n}1/${n}2/g *.c *.h
foreach? sed -i s/${n}0/${n}1/g *.c *.h
foreach? end

prompt> sed -i s/2-face/3-face/g *.c *.h
prompt> sed -i s/1-face/2-face/g *.c *.h
prompt> sed -i s/0-face/1-face/g *.c *.h


-- now that looks better, let's see if it compiles (before it didn't):
   -- compiled with a warning about "time()" being undefined in ranc.c
      -- changed the header references inside to include "time.h",
      -- compiles fine now

-- changing "makefile" to look more like that in harm-release
   -- seems to work

-- since none of the Num. Rec. stuff in nrutil.[c,h] is used, i'm
   removing the files
   -- compiles fine, so i guess I really didn't miss anything

-- committing these changes to CVS, eliminating nrutil.* from
   repository...


-- changed "makefile" to generate defs.h from decs.h ,
	  so decs.h will be where the mnenomics are defined
	  and where the global variables are added (using "extern ...")


#############################
Thu Sep 14 17:55:13 EDT 2006
-------------------------------

-- commenting and changing  decs.h
   -- adding grid index mnemonics based on N0-N3, e.g. N[0-3][S,E] for
      the start and ending indices for those directions,


-- changing L1,L2,L3 -> GridLength[0-3]:
   (made sure that nothing else matched on L[1-3] )

prompt> sed -i s/L1/GridLength\[1\]/g *.c *.h
prompt> sed -i s/L2/GridLength\[2\]/g *.c *.h
prompt> sed -i s/L3/GridLength\[3\]/g *.c *.h

    -- changed tf->GridLength[0]


-- need to change is->N1S , js->N2S, etc.:

-- changed is->N1s:
   -- checked for instances of "is" that was not the variable "is":
prompt> grep is *.c *.h | grep -v "[^[:alnum:]_]is[^[:alnum:]_]" >is.tmp

   -- made the replacement:
prompt> sed -i -e 's/is/N1S/g' *.c *.h

   -- then replaced those instances of "is" that were within words (from is.tmp)


   -- repeated this for js,ks,ie,je,ke
      (needed things like: )
prompt> sed -i -e 's/N2S_ad/js_ad/g' *.c *.h



-- compiles after these changes


-- change dt,dx1,dx2,dx3    ->  dX[0-3]

prompt> sed -i -e 's/dt/dX\[0\]/g' *.c *.h
prompt> sed -i -e 's/dX\[0\]sqijmin/dtsqijmin/g' *.c *.h
prompt> sed -i -e 's/dX\[0\]sqij/dtsqij/g' *.c *.h
prompt> sed -i -e 's/dX\[0\]t/dtt/g' *.c *.h
prompt> sed -i -e 's/dx1/dX\[1\]/g' *.c *.h
prompt> sed -i -e 's/dx2/dX\[2\]/g' *.c *.h
prompt> sed -i -e 's/dx3/dX\[3\]/g' *.c *.h



#############################
Fri Sep 15 10:51:47 EDT 2006
-------------------------------

-- changing dX[] -> dx[]  since it doesn't make sense to use uppercase
   since we typically write the contravariant line element as lower
   case;

-- since there isn't a need for a relativistic shearing box boundary
   condition, then I'm going to delete it; those needing it for the
   ham3d-mpi version will have to supply their own.
   -- then, i'm removing bounds_plin.c from the CVS



-- added "RECON_TYPE_CHOICE" macro and its options
-- added "BC_TYPE_CHOICE" macro and its options

-- changed bounds.c to add several different types of global
   boundary conditions (those of different coordinate systems)


#############################
Fri Sep 29 10:49:16 EDT 2006
-------------------------------

-- went through step_ch.c
   -- for now nothing to be done I believe
   -- may need to change wave_minmax() in phys.c
   -- will eventually need to change how flux is calculated
      to add different methods
   -- may need to eliminate timestep() calls to wave_minmax()
      since wave_minmax() calc. is already done in fluxcalc()

src/lu.c
/***********************************************************************************
    Copyright 2006 Charles F. Gammie, Jonathan C. McKinney, Scott C. Noble, 
                   Gabor Toth, and Luca Del Zanna

                        HARM  version 1.0   (released May 1, 2006)

    This file is part of HARM.  HARM is a program that solves hyperbolic 
    partial differential equations in conservative form using high-resolution
    shock-capturing techniques.  This version of HARM has been configured to 
    solve the relativistic magnetohydrodynamic equations of motion on a 
    stationary black hole spacetime in Kerr-Schild coordinates to evolve
    an accretion disk model. 

    You are morally obligated to cite the following two papers in his/her 
    scientific literature that results from use of any part of HARM:

    [1] Gammie, C. F., McKinney, J. C., \& Toth, G.\ 2003, 
        Astrophysical Journal, 589, 444.

    [2] Noble, S. C., Gammie, C. F., McKinney, J. C., \& Del Zanna, L. \ 2006, 
        Astrophysical Journal, 641, 626.

   
    Further, we strongly encourage you to obtain the latest version of 
    HARM directly from our distribution website:
    http://rainman.astro.uiuc.edu/codelib/


    HARM is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    HARM is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with HARM; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

***********************************************************************************/

#include "decs.h"

/*************************************************************************/
/*************************************************************************

   invert_matrix():

       Uses LU decomposition and back substitution to invert a matrix 
       A[][] and assigns the inverse to Ainv[][].  This routine does not 
       destroy the original matrix A[][].                                     
                                                                             
       Returns (1) if a singular matrix is found,  (0) otherwise.            

*************************************************************************/
int invert_matrix( double Am[][NDIM], double Aminv[][NDIM], double *detg )  
{ 

  int i,j;
  int n = NDIM;
  int permute[NDIM]; 
  double dxm[NDIM], Amtmp[NDIM][NDIM];
  int LU_decompose( double A[][NDIM], int permute[] );
  void LU_substitution( double A[][NDIM], double B[], int permute[] );
  double gdettmp;

  for( i = 0 ; i < NDIM*NDIM ; i++ ) {  Amtmp[0][i] = Am[0][i]; }

  // Get the LU matrix:
  if( LU_decompose( Amtmp,  permute ) != 0  ) { 
    fprintf(stderr, "invert_matrix(): singular matrix encountered! \n");
    return(1);
  }

  gdettmp = 1.;
  DLOOP1 gdettmp *= Amtmp[i][i];
  *detg = sqrt(fabs(gdettmp)) ;

  for( i = 0; i < n; i++ ) { 
    for( j = 0 ; j < n ; j++ ) { dxm[j] = 0. ; }
    dxm[i] = 1.; 
    
    /* Solve the linear system for the i^th column of the inverse matrix: :  */
    LU_substitution( Amtmp,  dxm, permute );

    for( j = 0 ; j < n ; j++ ) {  Aminv[j][i] = dxm[j]; }

  }

  return(0);
}

/*************************************************************************/
/*************************************************************************

   LU_decompose():

       Performs a LU decomposition of the matrix A using Crout's method      
       with partial implicit pivoting.  The exact LU decomposition of the    
       matrix can be reconstructed from the resultant row-permuted form via  
       the integer array permute[]                                            
                                                                             
       The algorithm closely follows ludcmp.c of "Numerical Recipes  
       in C" by Press et al. 1992.                                           
                                                                             
       This will be used to solve the linear system  A.x = B                 
                                                                             
       Returns (1) if a singular matrix is found,  (0) otherwise.            

*************************************************************************/
int LU_decompose( double A[][NDIM], int permute[] )
{

  const  double absmin = 1.e-30; /* Value used instead of 0 for singular matrices */

  static double row_norm[NDIM];
  double  absmax, maxtemp, mintemp;

  int i, j, k, max_row;


  max_row = 0;

  /* Find the maximum elements per row so that we can pretend later
     we have unit-normalized each equation: */

  for( i = 0; i < NDIM; i++ ) { 
    absmax = 0.;
    
    for( j = 0; j < NDIM ; j++ ) { 
      
      maxtemp = fabs( A[i][j] ); 

      if( maxtemp > absmax ) { 
	absmax = maxtemp; 
      }
    }

    /* Make sure that there is at least one non-zero element in this row: */
    if( absmax == 0. ) { 
     fprintf(stderr, "LU_decompose(): row-wise singular matrix!\n");
      return(1);
    }

    row_norm[i] = 1. / absmax ;   /* Set the row's normalization factor. */
  }


  /* The following the calculates the matrix composed of the sum 
     of the lower (L) tridagonal matrix and the upper (U) tridagonal
     matrix that, when multiplied, form the original maxtrix.  
     This is what we call the LU decomposition of the maxtrix. 
     It does this by a recursive procedure, starting from the 
     upper-left, proceding down the column, and then to the next
     column to the right.  The decomposition can be done in place 
     since element {i,j} require only those elements with {<=i,<=j} 
     which have already been computed.
     See pg. 43-46 of "Num. Rec." for a more thorough description. 
  */

  /* For each of the columns, starting from the left ... */
  for( j = 0; j < NDIM; j++ ) {

    /* For each of the rows starting from the top.... */

    /* Calculate the Upper part of the matrix:  i < j :   */
    for( i = 0; i < j; i++ ) {
      for( k = 0; k < i; k++ ) { 
	A[i][j] -= A[i][k] * A[k][j];
      }
    }

    absmax = 0.0;

    /* Calculate the Lower part of the matrix:  i <= j :   */

    for( i = j; i < NDIM; i++ ) {

      for (k = 0; k < j; k++) { 
	A[i][j] -= A[i][k] * A[k][j];
      }

      /* Find the maximum element in the column given the implicit 
	 unit-normalization (represented by row_norm[i]) of each row: 
      */
      maxtemp = fabs(A[i][j]) * row_norm[i] ;

      if( maxtemp >= absmax ) {
	absmax = maxtemp;
	max_row = i;
      }

    }

    /* Swap the row with the largest element (of column j) with row_j.  absmax
       This is the partial pivoting procedure that ensures we don't divide
       by 0 (or a small number) when we solve the linear system.  
       Also, since the procedure starts from left-right/top-bottom, 
       the pivot values are chosen from a pool involving all the elements 
       of column_j  in rows beneath row_j.  This ensures that 
       a row  is not permuted twice, which would mess things up. 
    */
    if( max_row != j ) {

      /* Don't swap if it will send a 0 to the last diagonal position. 
	 Note that the last column cannot pivot with any other row, 
	 so this is the last chance to ensure that the last two 
	 columns have non-zero diagonal elements.
       */

      if( (j == (NDIM-2)) && (A[j][j+1] == 0.) ) {
	max_row = j;
      }
      else { 
	for( k = 0; k < NDIM; k++ ) { 

	  maxtemp       = A[   j   ][k] ; 
	  A[   j   ][k] = A[max_row][k] ;
	  A[max_row][k] = maxtemp; 

	}

	/* Don't forget to swap the normalization factors, too... 
	   but we don't need the jth element any longer since we 
	   only look at rows beneath j from here on out. 
	*/
	row_norm[max_row] = row_norm[j] ; 
      }
    }

    /* Set the permutation record s.t. the j^th element equals the 
       index of the row swapped with the j^th row.  Note that since 
       this is being done in successive columns, the permutation
       vector records the successive permutations and therefore
       index of permute[] also indexes the chronology of the 
       permutations.  E.g. permute[2] = {2,1} is an identity 
       permutation, which cannot happen here though. 
    */

    permute[j] = max_row;

    if( A[j][j] == 0. ) { 
      A[j][j] = absmin;
    }


  /* Normalize the columns of the Lower tridiagonal part by their respective 
     diagonal element.  This is not done in the Upper part because the 
     Lower part's diagonal elements were set to 1, which can be done w/o 
     any loss of generality.
  */
    if( j != (NDIM-1) ) { 
      maxtemp = 1. / A[j][j]  ;
      
      for( i = (j+1) ; i < NDIM; i++ ) {
	A[i][j] *= maxtemp;
      }
    }

  }

  return(0);

  /* End of LU_decompose() */

}


/************************************************************************/
/************************************************************************

   LU_substitution():

       Performs the forward (w/ the Lower) and backward (w/ the Upper)   
       substitutions using the LU-decomposed matrix A[][] of the original
       matrix A' of the linear equation:  A'.x = B.  Upon entry, A[][]   
       is the LU matrix, B[] is the source vector, and permute[] is the  
       array containing order of permutations taken to the rows of the LU
       matrix.  See LU_decompose() for further details. 		     
    								     
       Upon exit, B[] contains the solution x[], A[][] is left unchanged.
								     
************************************************************************/
void LU_substitution( double A[][NDIM], double B[], int permute[] )
{
  int i, j ;
  double tmpvar,tmpvar2;

  
  /* Perform the forward substitution using the LU matrix. 
   */
  for(i = 0; i < NDIM; i++) {

    /* Before doing the substitution, we must first permute the 
       B vector to match the permutation of the LU matrix. 
       Since only the rows above the currrent one matter for 
       this row, we can permute one at a time. 
    */
    tmpvar        = B[permute[i]];
    B[permute[i]] = B[    i     ];
    for( j = (i-1); j >= 0 ; j-- ) { 
      tmpvar -=  A[i][j] * B[j];
    }
    B[i] = tmpvar; 
  }
	   

  /* Perform the backward substitution using the LU matrix. 
   */
  for( i = (NDIM-1); i >= 0; i-- ) { 
    for( j = (i+1); j < NDIM ; j++ ) { 
      B[i] -=  A[i][j] * B[j];
    }
    B[i] /= A[i][i] ; 
  }

  /* End of LU_substitution() */

}

#if( USE_STRICT_ARRAY_BOUNDS )

int invert_matrix2( double Am[][NDIM], double invOut[][NDIM], double *detg )  
{
  double inv[16], det;
  int i,j,k;

  inv[0] =   Am[1][1]*Am[2][2]*Am[3][3] - Am[1][1]*Am[2][3]*Am[3][2] - Am[2][1]*Am[1][2]*Am[3][3]
    + Am[2][1]*Am[1][3]*Am[3][2] + Am[3][1]*Am[1][2]*Am[2][3] - Am[3][1]*Am[1][3]*Am[2][2];
  inv[4] =  -Am[1][0]*Am[2][2]*Am[3][3] + Am[1][0]*Am[2][3]*Am[3][2] + Am[2][0]*Am[1][2]*Am[3][3]
    - Am[2][0]*Am[1][3]*Am[3][2] - Am[3][0]*Am[1][2]*Am[2][3] + Am[3][0]*Am[1][3]*Am[2][2];
  inv[8] =   Am[1][0]*Am[2][1]*Am[3][3] - Am[1][0]*Am[2][3]*Am[3][1] - Am[2][0]*Am[1][1]*Am[3][3]
    + Am[2][0]*Am[1][3]*Am[3][1] + Am[3][0]*Am[1][1]*Am[2][3] - Am[3][0]*Am[1][3]*Am[2][1];
  inv[12] = -Am[1][0]*Am[2][1]*Am[3][2] + Am[1][0]*Am[2][2]*Am[3][1] + Am[2][0]*Am[1][1]*Am[3][2]
    - Am[2][0]*Am[1][2]*Am[3][1] - Am[3][0]*Am[1][1]*Am[2][2] + Am[3][0]*Am[1][2]*Am[2][1];
  inv[1] =  -Am[0][1]*Am[2][2]*Am[3][3] + Am[0][1]*Am[2][3]*Am[3][2] + Am[2][1]*Am[0][2]*Am[3][3]
    - Am[2][1]*Am[0][3]*Am[3][2] - Am[3][1]*Am[0][2]*Am[2][3] + Am[3][1]*Am[0][3]*Am[2][2];
  inv[5] =   Am[0][0]*Am[2][2]*Am[3][3] - Am[0][0]*Am[2][3]*Am[3][2] - Am[2][0]*Am[0][2]*Am[3][3]
    + Am[2][0]*Am[0][3]*Am[3][2] + Am[3][0]*Am[0][2]*Am[2][3] - Am[3][0]*Am[0][3]*Am[2][2];
  inv[9] =  -Am[0][0]*Am[2][1]*Am[3][3] + Am[0][0]*Am[2][3]*Am[3][1] + Am[2][0]*Am[0][1]*Am[3][3]
    - Am[2][0]*Am[0][3]*Am[3][1] - Am[3][0]*Am[0][1]*Am[2][3] + Am[3][0]*Am[0][3]*Am[2][1];
  inv[13] =  Am[0][0]*Am[2][1]*Am[3][2] - Am[0][0]*Am[2][2]*Am[3][1] - Am[2][0]*Am[0][1]*Am[3][2]
    + Am[2][0]*Am[0][2]*Am[3][1] + Am[3][0]*Am[0][1]*Am[2][2] - Am[3][0]*Am[0][2]*Am[2][1];
  inv[2] =   Am[0][1]*Am[1][2]*Am[3][3] - Am[0][1]*Am[1][3]*Am[3][2] - Am[1][1]*Am[0][2]*Am[3][3]
    + Am[1][1]*Am[0][3]*Am[3][2] + Am[3][1]*Am[0][2]*Am[1][3] - Am[3][1]*Am[0][3]*Am[1][2];
  inv[6] =  -Am[0][0]*Am[1][2]*Am[3][3] + Am[0][0]*Am[1][3]*Am[3][2] + Am[1][0]*Am[0][2]*Am[3][3]
    - Am[1][0]*Am[0][3]*Am[3][2] - Am[3][0]*Am[0][2]*Am[1][3] + Am[3][0]*Am[0][3]*Am[1][2];
  inv[10] =  Am[0][0]*Am[1][1]*Am[3][3] - Am[0][0]*Am[1][3]*Am[3][1] - Am[1][0]*Am[0][1]*Am[3][3]
    + Am[1][0]*Am[0][3]*Am[3][1] + Am[3][0]*Am[0][1]*Am[1][3] - Am[3][0]*Am[0][3]*Am[1][1];
  inv[14] = -Am[0][0]*Am[1][1]*Am[3][2] + Am[0][0]*Am[1][2]*Am[3][1] + Am[1][0]*Am[0][1]*Am[3][2]
    - Am[1][0]*Am[0][2]*Am[3][1] - Am[3][0]*Am[0][1]*Am[1][2] + Am[3][0]*Am[0][2]*Am[1][1];
  inv[3] =  -Am[0][1]*Am[1][2]*Am[2][3] + Am[0][1]*Am[1][3]*Am[2][2] + Am[1][1]*Am[0][2]*Am[2][3]
    - Am[1][1]*Am[0][3]*Am[2][2] - Am[2][1]*Am[0][2]*Am[1][3] + Am[2][1]*Am[0][3]*Am[1][2];
  inv[7] =   Am[0][0]*Am[1][2]*Am[2][3] - Am[0][0]*Am[1][3]*Am[2][2] - Am[1][0]*Am[0][2]*Am[2][3]
    + Am[1][0]*Am[0][3]*Am[2][2] + Am[2][0]*Am[0][2]*Am[1][3] - Am[2][0]*Am[0][3]*Am[1][2];
  inv[11] = -Am[0][0]*Am[1][1]*Am[2][3] + Am[0][0]*Am[1][3]*Am[2][1] + Am[1][0]*Am[0][1]*Am[2][3]
    - Am[1][0]*Am[0][3]*Am[2][1] - Am[2][0]*Am[0][1]*Am[1][3] + Am[2][0]*Am[0][3]*Am[1][1];
  inv[15] =  Am[0][0]*Am[1][1]*Am[2][2] - Am[0][0]*Am[1][2]*Am[2][1] - Am[1][0]*Am[0][1]*Am[2][2]
    + Am[1][0]*Am[0][2]*Am[2][1] + Am[2][0]*Am[0][1]*Am[1][2] - Am[2][0]*Am[0][2]*Am[1][1];

  det = Am[0][0]*inv[0] + Am[0][1]*inv[4] + Am[0][2]*inv[8] + Am[0][3]*inv[12];

  if( det == 0 ) {
    struct of_coord *coords;
    get_coord(icurr,jcurr,kcurr,pcurr,ncurr,coords);
    fprintf(stderr, "invert_matrix2(): singular matrix encountered!  at  ijk=[%d, %d, %d]  pos=%d  t,x,y,z = (%26.16e,%26.16e,%26.16e,%26.16e) \n",
	    icurr,jcurr,kcurr,pcurr, t,coords->xcart[1],coords->xcart[2],coords->xcart[3]); 
    fflush(stderr); 
    return(1);
  }

  *detg = sqrt(fabs(det));

  det = 1.0 / det;

  k = 0; 
  DLOOP2 { invOut[i][j] = inv[k++] * det; }

  return(0); 

}

#else 

int invert_matrix2( double Am[][NDIM], double invOut[][NDIM], double *detg )  
{
  double inv[16], det;
  int i;

  inv[0] =   Am[0][5]*Am[0][10]*Am[0][15] - Am[0][5]*Am[0][11]*Am[0][14] - Am[0][9]*Am[0][6]*Am[0][15]
    + Am[0][9]*Am[0][7]*Am[0][14] + Am[0][13]*Am[0][6]*Am[0][11] - Am[0][13]*Am[0][7]*Am[0][10];
  inv[4] =  -Am[0][4]*Am[0][10]*Am[0][15] + Am[0][4]*Am[0][11]*Am[0][14] + Am[0][8]*Am[0][6]*Am[0][15]
    - Am[0][8]*Am[0][7]*Am[0][14] - Am[0][12]*Am[0][6]*Am[0][11] + Am[0][12]*Am[0][7]*Am[0][10];
  inv[8] =   Am[0][4]*Am[0][9]*Am[0][15] - Am[0][4]*Am[0][11]*Am[0][13] - Am[0][8]*Am[0][5]*Am[0][15]
    + Am[0][8]*Am[0][7]*Am[0][13] + Am[0][12]*Am[0][5]*Am[0][11] - Am[0][12]*Am[0][7]*Am[0][9];
  inv[12] = -Am[0][4]*Am[0][9]*Am[0][14] + Am[0][4]*Am[0][10]*Am[0][13] + Am[0][8]*Am[0][5]*Am[0][14]
    - Am[0][8]*Am[0][6]*Am[0][13] - Am[0][12]*Am[0][5]*Am[0][10] + Am[0][12]*Am[0][6]*Am[0][9];
  inv[1] =  -Am[0][1]*Am[0][10]*Am[0][15] + Am[0][1]*Am[0][11]*Am[0][14] + Am[0][9]*Am[0][2]*Am[0][15]
    - Am[0][9]*Am[0][3]*Am[0][14] - Am[0][13]*Am[0][2]*Am[0][11] + Am[0][13]*Am[0][3]*Am[0][10];
  inv[5] =   Am[0][0]*Am[0][10]*Am[0][15] - Am[0][0]*Am[0][11]*Am[0][14] - Am[0][8]*Am[0][2]*Am[0][15]
    + Am[0][8]*Am[0][3]*Am[0][14] + Am[0][12]*Am[0][2]*Am[0][11] - Am[0][12]*Am[0][3]*Am[0][10];
  inv[9] =  -Am[0][0]*Am[0][9]*Am[0][15] + Am[0][0]*Am[0][11]*Am[0][13] + Am[0][8]*Am[0][1]*Am[0][15]
    - Am[0][8]*Am[0][3]*Am[0][13] - Am[0][12]*Am[0][1]*Am[0][11] + Am[0][12]*Am[0][3]*Am[0][9];
  inv[13] =  Am[0][0]*Am[0][9]*Am[0][14] - Am[0][0]*Am[0][10]*Am[0][13] - Am[0][8]*Am[0][1]*Am[0][14]
    + Am[0][8]*Am[0][2]*Am[0][13] + Am[0][12]*Am[0][1]*Am[0][10] - Am[0][12]*Am[0][2]*Am[0][9];
  inv[2] =   Am[0][1]*Am[0][6]*Am[0][15] - Am[0][1]*Am[0][7]*Am[0][14] - Am[0][5]*Am[0][2]*Am[0][15]
    + Am[0][5]*Am[0][3]*Am[0][14] + Am[0][13]*Am[0][2]*Am[0][7] - Am[0][13]*Am[0][3]*Am[0][6];
  inv[6] =  -Am[0][0]*Am[0][6]*Am[0][15] + Am[0][0]*Am[0][7]*Am[0][14] + Am[0][4]*Am[0][2]*Am[0][15]
    - Am[0][4]*Am[0][3]*Am[0][14] - Am[0][12]*Am[0][2]*Am[0][7] + Am[0][12]*Am[0][3]*Am[0][6];
  inv[10] =  Am[0][0]*Am[0][5]*Am[0][15] - Am[0][0]*Am[0][7]*Am[0][13] - Am[0][4]*Am[0][1]*Am[0][15]
    + Am[0][4]*Am[0][3]*Am[0][13] + Am[0][12]*Am[0][1]*Am[0][7] - Am[0][12]*Am[0][3]*Am[0][5];
  inv[14] = -Am[0][0]*Am[0][5]*Am[0][14] + Am[0][0]*Am[0][6]*Am[0][13] + Am[0][4]*Am[0][1]*Am[0][14]
    - Am[0][4]*Am[0][2]*Am[0][13] - Am[0][12]*Am[0][1]*Am[0][6] + Am[0][12]*Am[0][2]*Am[0][5];
  inv[3] =  -Am[0][1]*Am[0][6]*Am[0][11] + Am[0][1]*Am[0][7]*Am[0][10] + Am[0][5]*Am[0][2]*Am[0][11]
    - Am[0][5]*Am[0][3]*Am[0][10] - Am[0][9]*Am[0][2]*Am[0][7] + Am[0][9]*Am[0][3]*Am[0][6];
  inv[7] =   Am[0][0]*Am[0][6]*Am[0][11] - Am[0][0]*Am[0][7]*Am[0][10] - Am[0][4]*Am[0][2]*Am[0][11]
    + Am[0][4]*Am[0][3]*Am[0][10] + Am[0][8]*Am[0][2]*Am[0][7] - Am[0][8]*Am[0][3]*Am[0][6];
  inv[11] = -Am[0][0]*Am[0][5]*Am[0][11] + Am[0][0]*Am[0][7]*Am[0][9] + Am[0][4]*Am[0][1]*Am[0][11]
    - Am[0][4]*Am[0][3]*Am[0][9] - Am[0][8]*Am[0][1]*Am[0][7] + Am[0][8]*Am[0][3]*Am[0][5];
  inv[15] =  Am[0][0]*Am[0][5]*Am[0][10] - Am[0][0]*Am[0][6]*Am[0][9] - Am[0][4]*Am[0][1]*Am[0][10]
    + Am[0][4]*Am[0][2]*Am[0][9] + Am[0][8]*Am[0][1]*Am[0][6] - Am[0][8]*Am[0][2]*Am[0][5];

  det = Am[0][0]*inv[0] + Am[0][1]*inv[4] + Am[0][2]*inv[8] + Am[0][3]*inv[12];

  if( det == 0 ) {
    struct of_coord *coords;
    get_coord(icurr,jcurr,kcurr,pcurr,ncurr,coords);
    fprintf(stderr, "invert_matrix2(): singular matrix encountered!  at  ijk=[%d, %d, %d]  pos=%d  t,x,y,z = (%26.16e,%26.16e,%26.16e,%26.16e) \n",
	    icurr,jcurr,kcurr,pcurr, t,coords->xcart[1],coords->xcart[2],coords->xcart[3]); 
    fflush(stderr); 
    return(1);
  }

  *detg = sqrt(fabs(det));

  det = 1.0 / det;

  for (i = 0; i < 16; i++)
    invOut[0][i] = inv[i] * det;

  return(0); 

}

#endif  /*  USE_STRICT_ARRAY_BOUNDS  */
